### ISSUE TITLE
Add support for defining COM interface wrappers in Go via structs and struct tags

### ISSUE THREAD

**Issue by mgspross** [mattn/go-ole#43]
I am playing around with the idea of adding support for defining COM interfaces in Go using struct's to define the interface, where each field is actually the definition of a property or method, using `func` types to specify the method signature, and struct tags to specify whether the field represents a method or a property on the COM object.

This is easier to show than to explain, so here is an example of how `MSXML2.DomDocument`'s interface would be (partially) implemented using this technique:

```
// COM "interface" definition for MSXML2.DomDocument.
//
// Each method/property is declared with a field with the same name as the method or property,
// and the field's type is set to a function type that matches the method/property's call signature (using Go types)
//
// The `ole` tag is used to specify whether a given field represents a method or a property for objects of this type
//
type DomDocument struct {
	Load    func(file string) bool `ole:"method"`
	LoadXml func(xml string) bool  `ole:"method"`
	Xml     func() string          `ole:"property"`
	Save    func(file string)      `ole:"method"`
}
```

You would use it as follows:

```
doc := DomDocument{}
oleutil.CreateOleWrapper("MSXML.DomDocument.6.0", &doc)

if(doc.LoadXml("<root><child/></root>")) {
    fmt.Println(doc.Xml())
}
```

This works by using reflection to find fields in the struct that have a function type, and dynamically creating wrapper functions that call `oleutil.MustCallMethod` or `oleutil.MustGetProperty` as needed.

For now, this is just a little experiment, and we'd have to flesh out a number of things (like returning errors from methods, properly releasing objects that are created, etc.).

I wrote a complete, working proof-of-concept here:

https://gist.github.com/mgspross/dfadc21f16db2a56ed27

What do you guys think of the general concept here?

**Comment by jacobsantos** [comment]
@mgspross I am thinking about working on that as I write the C# COM Server for testing and possible examples. Need to have a way or reason to do it all. I'm not sure how the CLR will handle all of the supported types. If not, then might need to break into C++ and write a COM server using that as well. There do seem to be more C++ COM examples and books than for .NET.

----------

One suggestion would be to attempt to marshal and unmarshal to follow the Golang naming convention.

Something else that could be useful is going off of gobind tool and writing the Go code based on the struct(s) in the project. So they wouldn't even need to write the COM code. What would be even more useful is if the didn't even need to write the structs. The tool would simply accept the application name or GUID and create the files and code needed to where they would simply access the Go object like they would any other Go object.

Second idea, might need to be a separate project, but I'm hoping the addition to gobind tool could be useful enough for inclusion instead of needing another tool and project.

----------

There are times with Go that I wish allowed for capture methods for missing methods and properties to better allow for dynamic access.

**Comment by jacobsantos** [comment]
@mgspross Is there something that you need to support this better? I think the tool to generate Go code is a good way to do this, but I'm up for any solution that is eloquent as this one seems to be. I'm working on the code now, so any suggestions would be better. I have implemented support for more Variant types, but I'm refactoring that system to I hope what would be a better.

Having to support every possible type is not a good system, I think. Well, for Variant types, I have already added support for majority of the types that are capable of being supported by Variant. I think I'm missing `VT_DATE`. I think I also added support for something that Variant couldn't be, but that can be corrected later.

**Comment by mgspross** [comment]
I'll try to break up my reply to address the different topics that have come up so far:

### This approach vs. the `go generate` approach

> **TL;DR**: If we pursue the `go generate` approach, implementing the approach in this issue is not really necessary, unless someone can come up with a reason why this would be useful in addition to the `go generate` approach.

@jacobsantos The idea presented here and the other proposal to create a tool to generate Go code for COM types are both essentially meant to solve the same problem - that is, providing a nice layer on top of `go-ole` to work with, so that it's more natural/easier to use COM in Go.

I thought of this idea before the idea to create a code generator was brought up. I'm not sure if my idea here (creating COM wrappers dynamically with structs and reflection) would be useful and/or necessary if we are going to move ahead with the `go generate` approach.

The `go generate` approach does seem much cleaner to me, and possibly simpler to implement - in the approach I outlined in this issue, I tried to make the API as clean as possible, but it's still (for various reasons) not possible to make a completely-natural Go-like API (for example, having to use `func` types in the so-called "interface" structs, instead of having "normal" Go functions that take a receiver).

So, I'm fine with not worrying about this particular feature and moving ahead with the `go generate` approach. I'd say my approach here is only useful if someone else can come up with a good reason to implement it and a rationale for why it would still be useful in light of also supporting the `go generate` approach.

### Support for more data types

As far as supporting more data types, my reason for wanting them was for better compatibility with different COM libraries. In my case, it was also to allow me to use `go-ole` with COM libraries written in VB6, and the API's for those COM libraries happened to use data types that aren't currently supported by `go-ole` in some cases.

For example, I had a few functions in one COM class I needed to work with that accepted and returned byte arrays, which is why I added support for converting Go byte slices to and from COM byte arrays awhile back.

We already support a decent number of types, and I'd have to look again to see if we're actually missing any right now. I know at one point some of the types that we could convert from Go to COM couldn't be converted from COM back to Go, because the functions to do the opposite conversion simply hadn't been implemented to support going in both directions for all the types yet. 

In addition, for `SAFEARRAY`'s, we weren't supporting all the possible `SAFEARRAY` types: basically, any of the types we currently support can also be used in a `SAFEARRAY`, so for consistency, I thought it would be useful to ensure that `SAFEARRAY` types in `go-ole` supported the same data types `go-ole` supports in general (i.e. to support having a `SAFEARRAY` of `int32`'s, `int64`'s, and so on).

### A note on `VT_DATE`

`VT_DATE` is an interesting type. It's my understanding that the underlying Windows COM subsystem, at least for `IDispatch:Invoke` calls, can accept regular strings for `VT_DATE` parameters, and it will do the necessary conversion to an actual date type. I've used some COM functions written in VB6 that took `Date` parameters, but I was able to use regular Go strings when I called the functions from `go-ole` by just passing in correctly-formatted date/time strings.

I'm not sure if this helps, but I thought it was interesting.

To actually implement "true" `VT_DATE` support, we'd need to support conversion between `time.Time` structs and the native COM/OLE datetime format, which is an odd format - in COM/OLE, datetimes are stored as double-precision numbers, with the integer part representing the day (relative to a specific "epoch" time) and the fractional part representing the time (relative to the start of the day) - there are some interesting edges case to deal with too.

Here's a quick blog post about some of the weirdness with the way COM/OLE handles datetimes: [Eric's Complete Guide To VT_DATE](http://blogs.msdn.com/b/ericlippert/archive/2003/09/16/eric-s-complete-guide-to-vt-date.aspx)

I'm not sure if we'd have to deal with the raw floating-point values in Go and convert it to the correct Go date/time (and vice-versa) ourselves, or if there might be some helper functions in the Windows API to help with this.

But, again, maybe we can get away with just letting the automagic string conversion that seems to happen handle `VT_DATE` for us.

**Comment by jacobsantos** [comment]
Okay. I am going to close this then, because I think the go generate will allow for much more features as well as cleaner, up to date solutions as the library evolves.

The tool will evolve. The first code I am going to generate is interface code matching the style that exists for IUnknown. I am not sure what style I am going to use for IDispatch Invoke calls on functions that take a receiver. I think it will improve as people use it and determine the best output for the code generator.

----------

I did do some research into the various COM types. I understand that the currency was probably before the IEEE standard and probably wanted to ensure precision with money. The date is interesting architecture choice. I am sure they thought it was good at the time.

I am sure we can talk more about the Date type in a separate ticket. I was just thinking of using Windows API functions to convert instead. Handling it directly seems like it would be a nightmare, when dealing with leap years and calendars.

I think it is great that Windows does that, but I don't think receiving does that and I think that is the part that needs implementing.

-------

One of my plans is fully supporting all types for SafeArrays. My plan is to simply complete the library or at least, however much that is possible for Go.

The current pattern I think for how it is implemented doesn't seem to be good enough in some cases. I don't like that you have to edit the library any time you want to add support for a new type. This might simply be ignorance of mine, but it seems like in one part, you have to modify the invoke IDispatch method and I would like to avoid that.

I am still working on this part, so I hope the solution I create works well enough. It might be over architecting a solution, but it can always be changed in the future.

---------

I hope to actually start on the go generate code in a week or two. I broke all the code with the restructuring of directories and files. I need to get it working again and also implement the itypelib interface.

It is coming together, but I am trying to make it where it supports more than what it does now. Also be more feature complete.
